rfcs/0013-ai-policies.md

```diff
@@ -0,0 +1,486 @@
+metadata:
+  name: token-limit-free
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: my-llm-gateway
+  limit:
+    rate:
+      limit: 20000
+      window: 1d
+    predicate: 'request.auth.claims["kuadrant.io/groups"].split(",").exists(g, g == "free")' 
+    counter: auth.identity.userid
+---
+apiVersion: kuadrant.io/v1alpha1
+kind: LLMPromptRiskCheckPolicy
+metadata:
+  name: prompt-check
+spec:
+  model:
+    url: http://huggingface-granite-guardian-default.example.com/v1
+    apiKey:
+      secretRef:
+        name: granite-api-key
+        key: token
+  categories:
+    - harm
+    - violence
+    - sexual_content
+  response:
+    unauthorized:
+      headers:
+        "content-type":
+          value: application/json
+      body:
+        value: |
+          {
+            "error": "Unauthorized",
+            "message": "Request prompt blocked by content policy."
+          }
+---
+apiVersion: kuadrant.io/v1alpha1
+kind: LLMResponseRiskCheckPolicy
+metadata:
+  name: completion-check
+spec:
+  model:
+    url: http://huggingface-granite-guardian-default.example.com/v1
+    apiKey:
+      secretRef:
+        name: granite-api-key
+        key: token
+  categories:
+    - harm
+  response:
+    forbidden:
+      headers:
+        "content-type":
+          value: application/json
+      body:
+        value: |
+          {
+            "error": "Forbidden",
+            "message": "Response blocked by content policy."
+          }
+```
+
+## User Impact
+
+- Existing users: These policies extend existing workflows. Policy attachment is the same; what's different is the semantics (tokens vs requests, prompt filtering, etc).
+- New users: Can adopt Kuadrant to secure and manage LLM workloads from the start.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- How error would be reported to the users.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+
+## `LLMTokenRateLimitPolicy`
+
+- Add a new resource, `LLMTokenRateLimitPolicy`, to be managed by the Kuadrant operator.
+- Either:
+  - Extend our existing `wasm-shim` to optionally amend the existing `actionSet` to optionally call both the guard filter and the token parsing filter implementation.
+  - Create a new `ext_proc` gRPC service for parsing OpenAI-style and usage metrics and adding these as well-known dynamic metadata, for use by Limitador
+- Extend the wasm-shim and `RateLimitPolicy` to give a means to specify an increment (currently, [hard-coded](https://github.com/Kuadrant/wasm-shim/blob/main/src/service/rate_limit.rs#L18) to `1`)
```

Comment:
I don't know if this is needed in the RLP api? it seems a very special use case. I wonder instead if it could be made some form of internal config not exposed to the user at this point?

Reply:
that is an option yes: we either inc the counter in a custom fashion in something in this filter chain, or we extend RLP somehow to support. if there's general utility in RLP, I guess that route may be more preferable

slight worry about having another mechanism other than limitador doing it - would end up needing to re-implment/copy a bunch of existing machinery? unsure

Reply:
hmm I might be misunderstanding. I was instead thinking of using the `hits_addend` and setting it dynamically if it is an AI interaction. We need @eguzki or @alexsnaps here interested to know what there thoughts are on how to send a custom increment.

Reply:
While this isn't implemented afaik, [well-known attributes](https://github.com/Kuadrant/architecture/blob/main/rfcs/0002-well-known-attributes.md#rate-limit-attributes) were meant to support this. So one way would be to have `ratelimit.hits_addend` (while defaulting to `1`) be mutable by "upstream" actions. So that a policy could set it to some arbitrary value _before_ the request to limitador is made.

Reply:
and, yes! 🙌 , `service.ext_proc.v3.ProcessingResponse` support [dynamic metadata](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ext_proc/v3/external_processor.proto#envoy-v3-api-field-service-ext-proc-v3-processingresponse-dynamic-metadata) to have data flow between envoy and these processes. If wasm-shim dispatches the call, then no worries in using that for that. If envoy does, we have to check what happens to it and how/if we can read them back properly from wasm tho (should be fine™ - t&c apply)